xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -0,0 +1,173 @@
+from collections.abc import Sequence
+
+from attr import dataclass
+
+from xdsl.context import MLContext
+from xdsl.dialects import memref, stencil, tensor
+from xdsl.dialects.builtin import ModuleOp, TensorType
+from xdsl.dialects.csl import csl_stencil
+from xdsl.dialects.experimental import dmp
+from xdsl.ir import Attribute, OpResult
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.utils.hints import isa
+
+
+@dataclass(frozen=True)
+class AccessOpFromPrefetchPattern(RewritePattern):
+    """
+    Rebuilds stencil.access by csl_stencil.access which operates on prefetched accesses.
+
+    stencil.access operates on stencil.temp types found at arg_index
+    csl_stencil.access operates on memref< num_neighbors x tensor< buf_size x data_type >> found at last arg index
+
+    Note: This is intended to be called in a nested pattern rewriter, such that the above precondition is met.
+    """
+
+    arg_index: int
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: stencil.AccessOp, rewriter: PatternRewriter, /):
+        assert len(op.offset) == 2
+        if tuple(op.offset) == (0, 0):
+            return
+        prefetched_arg = op.get_apply().region.block.args[-1]
+        assert isa(m_type := prefetched_arg.type, memref.MemRefType[Attribute])
+        assert isa(t_type := m_type.get_element_type(), TensorType[Attribute])
+
+        csl_access_op = csl_stencil.AccessOp(
+            op=prefetched_arg,
+            offset=op.offset,
+            offset_mapping=op.offset_mapping,
+            result_type=t_type,
+        )
+
+        # The stencil-tensorize-z-dimension pass inserts tensor.ExtractSliceOps after stencil.access to remove ghost cells.
+        # Since ghost cells are not prefetched, these ops can be removed again. Check if the ExtractSliceOp
+        # has no other effect and if so, remove both.
+        if (
+            len(op.res.uses) == 1
+            and isinstance(use := list(op.res.uses)[0].operation, tensor.ExtractSliceOp)
+            and tuple(d.data for d in use.static_sizes.data) == t_type.get_shape()
+            and tuple(d.data for d in use.static_offsets.data) == (0,)
+            and tuple(d.data for d in use.static_strides.data) == (1,)
+            and len(use.offsets) == 0
+            and len(use.sizes) == 0
+            and len(use.strides) == 0
+        ):
+            rewriter.replace_op(use, csl_access_op)
+            rewriter.erase_op(op)
+        else:
+            rewriter.replace_matched_op(csl_access_op)
```

Comment:
This feels like it should be a canonicalization pattern on the `tensor.ExtractSliceOp` instead of part of this pattern, or am I missing something?

Reply:
I strongly prefer to finish the dialect and transform first, as the design doc suggests there will be some follow-up rewriting, for which I'd prefer not to have the extract slice ops if at all possible.